impl/src/test/groovy/com/okta/sdk/impl/cache/DefaultCacheTest.groovy

```diff
@@ -0,0 +1,266 @@
+    void testSetZeroTti() {
+        def cache = new DefaultCache('foo')
+        def tti = Duration.ofMillis(0)
+        cache.setTimeToIdle(tti)
+    }
+
+    @Test
+    void testRemove() {
+        def cache = new DefaultCache('foo')
+        def existing = cache.put('key', 'value')
+        assertNull existing
+
+        existing = cache.remove('key')
+        assertEquals existing, 'value'
+
+        assertEquals cache.accessCount, 1
+        assertEquals cache.hitCount, 1
+        assertEquals cache.missCount, 0
+        assertEquals cache.hitRatio, 1.0d
+
+        def value = cache.get('key')
+        assertNull value
+
+        assertEquals cache.accessCount, 2
+        assertEquals cache.hitCount, 1
+        assertEquals cache.missCount, 1
+        def ratio = cache.hitRatio
+        assertEquals ratio, 0.5d
+
+        value = cache.remove('key')
+        assertNull value
+
+        assertEquals cache.accessCount, 3
+        assertEquals cache.hitCount, 1
+        assertEquals cache.missCount, 2
+        ratio = cache.hitRatio
+        assertEquals ratio, (1d/3d)
+    }
+
+
+    @Test
+    void testClear() {
+        def cache = new DefaultCache('foo')
+        cache.put('key', 'value')
+        assertEquals cache.size(), 1
+        cache.clear()
+        assertEquals cache.size(), 0
+    }
+
+    @Test
+    void testToString() {
+        def cache = new DefaultCache('foo')
+        def json = new JsonSlurper().parseText(cache.toString())
+
+        assertEquals json.name, 'foo'
+        assertEquals json.size, 0
+        assertEquals json.accessCount, 0
+        assertEquals json.hitCount, 0
+        assertEquals json.missCount, 0
+        assertEquals json.hitRatio, 0.0
+
+        cache.put('key', 'value')
+        def value = cache.get('key')
+        assertEquals value, 'value'
+
+        json = new JsonSlurper().parseText(cache.toString())
+
+        assertEquals json.name, 'foo'
+        assertEquals json.size, 1
+        assertEquals json.accessCount, 1
+        assertEquals json.hitCount, 1
+        assertEquals json.missCount, 0
+        assertEquals json.hitRatio, 1.0
+    }
+
+    @Test
+    void testPutReplacesPreviousValue() {
+        def cache = new DefaultCache('foo', [:], null, null);
+
+        def key = 'key'
+        def value1 = 'value1'
+        def value2 = 'value2'
+
+        assertNull cache.put(key, value1)
+
+        def prev = cache.put(key, value2)
+
+        assertEquals prev, value1
+        assertEquals cache.get(key), value2
+    }
+
+    @Test(enabled = false) // TODO: flaky test, need to fix
+    void testTimeToLive() {
+
+        def cache = new DefaultCache('foo', [:], Duration.ofMillis(10), null)
+
+        def key = 'key'
+        def value = 'value'
+
+        def prev = cache.put(key, value)
+        assertNull prev
+
+        def found = cache.get(key)
+        assertEquals value, found
+        assertEquals cache.size(), 1
+
+        Thread.sleep(15)
```

Comment:
Can we setup the cache with a `Clock` interface, that way we can set the time in tests, rather than using real time/sleeps? These tests don't scale well, as the test runtime will increase heavily as new tests are added.

Reply:
Sure, that'd be a good idea. Like I mentioned earlier, we wanted to port the existing Cache impl from the v2 SDK to make it less effort intensive for the moment. But, your idea does make sense and will definitely consider that while doing future improvements.